File: test/python-tests/test_copy_pickle_shelve.py
#!/usr/bin/env python3
"""
测试Python的深拷贝vs浅拷贝行为
"""

import copy
import pickle
import shelve
import os
import tempfile

def test_shallow_vs_deep_copy():
    """测试浅拷贝和深拷贝的区别"""
    print("=== Testing Shallow vs Deep Copy ===")
    
    # 创建嵌套数据结构
    original = {
        'numbers': [1, 2, 3],
        'nested': {'inner': [4, 5, 6]},
        'simple': 'string'
    }
    
    # 浅拷贝
    shallow = copy.copy(original)
    
    # 深拷贝
    deep = copy.deepcopy(original)
    
    print("Original:", original)
    print("Shallow copy:", shallow)
    print("Deep copy:", deep)
    
    # 修改原始数据
    original['numbers'].append(999)
    original['nested']['inner'].append(888)
    original['simple'] = 'modified'
    
    print("\nAfter modifying original:")
    print("Original:", original)
    print("Shallow copy (affected by nested changes):", shallow)
    print("Deep copy (unaffected):", deep)

def test_copy_edge_cases():
    """测试拷贝的边缘情况"""
    print("\n=== Testing Copy Edge Cases ===")
    
    # 测试循环引用
    a = [1, 2, 3]
    b = [4, 5, 6]
    a.append(b)
    b.append(a)  # 创建循环引用
    
    print("Circular reference before copy:")
    print(f"a: {a}")
    print(f"b: {b}")
    
    # 深拷贝循环引用
    a_copy = copy.deepcopy(a)
    print("\nDeep copy of circular reference:")
    print(f"a_copy: {a_copy}")
    
    # 修改原始数据验证独立性
    a[0] = 999
    print(f"\nAfter modifying original a[0] = 999:")
    print(f"original a: {a}")
    print(f"copied a_copy: {a_copy}")

def test_pickle_serialization():
    """测试pickle序列化"""
    print("\n=== Testing Pickle Serialization ===")
    
    # 创建测试数据
    data = {
        'string': 'hello world',
        'list': [1, 2.5, True, None],
        'dict': {'nested': {'deep': 'value'}},
        'tuple': (1, 2, 3),
        'set': {4, 5, 6}
    }
    
    print("Original data:", data)
    
    # 序列化
    pickled = pickle.dumps(data)
    print(f"Pickled data size: {len(pickled)} bytes")
    
    # 反序列化
    unpickled = pickle.loads(pickled)
    print("Unpickled data:", unpickled)
    
    # 验证数据相等性
    print("Data equality:", data == unpickled)
    
    # 测试不同协议版本
    for protocol in range(pickle.HIGHEST_PROTOCOL + 1):
        try:
            p = pickle.dumps(data, protocol=protocol)
            u = pickle.loads(p)
            print(f"Protocol {protocol}: size = {len(p)} bytes, works = {data == u}")
        except Exception as e:
            print(f"Protocol {protocol}: failed - {e}")

def test_shelve_persistence():
    """测试shelve持久化存储"""
    print("\n=== Testing Shelve Persistence ===")
    
    # 创建临时数据库文件（使用简单的文件名，避免扩展名问题）
    temp_dir = tempfile.mkdtemp()
    db_path = os.path.join(temp_dir, 'test_shelve')
    
    try:
        # 写入数据到shelve数据库
        with shelve.open(db_path) as db:
            db['user_data'] = {'name': 'Alice', 'age': 30}
            db['settings'] = {'theme': 'dark', 'language': 'en'}
            db['numbers'] = [1, 2, 3, 4, 5]
            db['timestamp'] = time.time()
            print("Data written to shelve database")
        
        # 从shelve数据库读取数据
        with shelve.open(db_path) as db:
            print("Data read from shelve database:")
            for key, value in db.items():
                print(f"  {key}: {value}")
        
        # 修改数据
        with shelve.open(db_path) as db:
            user_data = db['user_data']
            user_data['age'] = 31  # 修改年龄
            db['user_data'] = user_data  # 写回数据库
            db['new_key'] = 'new_value'  # 添加新键
            print("\nData modified in database")
        
        # 验证修改
        with shelve.open(db_path) as db:
            print("Modified data:")
            print(f"  user_data: {db['user_data']}")
            print(f"  new_key: {db['new_key']}")
    
    finally:
        # 清理临时文件和目录
        import glob
        for file_path in glob.glob(f"{db_path}*"):
            try:
                os.unlink(file_path)
            except:
                pass
        try:
            os.rmdir(temp_dir)
        except:
            pass

import time

def test_copy_performance():
    """测试拷贝性能"""
    print("\n=== Testing Copy Performance ===")
    
    # 创建大数据结构
    large_data = {
        'items': [{'id': i, 'data': list(range(100))} for i in range(1000)]
    }
    
    # 测试浅拷贝性能
    start = time.time()
    shallow_copy = copy.copy(large_data)
    shallow_time = time.time() - start
    
    # 测试深拷贝性能
    start = time.time()
    deep_copy = copy.deepcopy(large_data)
    deep_time = time.time() - start
    
    print(f"Shallow copy time: {shallow_time:.4f} seconds")
    print(f"Deep copy time: {deep_time:.4f} seconds")
    print(f"Deep copy is {deep_time/shallow_time:.1f}x slower")
    
    # 验证拷贝的正确性
    print(f"Shallow copy first item data == original: {shallow_copy['items'][0]['data'] == large_data['items'][0]['data']}")
    print(f"Deep copy first item data == original: {deep_copy['items'][0]['data'] == large_data['items'][0]['data']}")
    
    # 修改验证独立性
    large_data['items'][0]['data'][0] = 99999
    print(f"After modifying original, shallow copy reflects change: {shallow_copy['items'][0]['data'][0] == 99999}")
    print(f"After modifying original, deep copy remains unchanged: {deep_copy['items'][0]['data'][0] == 0}")

if __name__ == "__main__":
    test_shallow_vs_deep_copy()
    test_copy_edge_cases()
    test_pickle_serialization()
    test_shelve_persistence()
    test_copy_performance()
    print("\n=== All copy and serialization tests completed ===")

File: test/python-tests/test_copy_pickle_shelve.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


void test_shallow_vs_deep_copy() {
    /* docstring */
    std::cout << "=== Testing Shallow vs Deep Copy ===" << std::endl;
    auto original = {;
        'numbers': std::vector<decltype(1)>{1, 2, 3},;
        'nested': {'inner': std::vector<decltype(4)>{4, 5, 6}},;
        'simple': 'string';
    }
    };
    auto shallow = copy.copy(original);
    auto deep = copy.deepcopy(original);
    std::cout << "Original:" << " " << original << std::endl;
    std::cout << "Shallow copy:" << " " << shallow << std::endl;
    std::cout << "Deep copy:" << " " << deep << std::endl;
    original['numbers'].push_back(999);
    original['nested']['inner'].push_back(888);
    auto original['simple'] = 'modified';
    std::cout << "\nAfter modifying original:" << std::endl;
    std::cout << "Original:" << " " << original << std::endl;
    std::cout << "Shallow copy (affected by nested changes):" << " " << shallow << std::endl;
    std::cout << "Deep copy (unaffected):" << " " << deep << std::endl;
}


void test_copy_edge_cases() {
    /* docstring */
    std::cout << "\n=== Testing Copy Edge Cases ===" << std::endl;
    auto a = std::vector<decltype(1)>{1, 2, 3};
    auto b = std::vector<decltype(4)>{4, 5, 6};
    a.push_back(b);
    b.push_back(a);
    std::cout << "Circular reference before copy:" << std::endl;
    std::cout << to_str("a: ") << to_str(a) << std::endl;
    std::cout << to_str("b: ") << to_str(b) << std::endl;
    auto a_copy = copy.deepcopy(a);
    std::cout << "\nDeep copy of circular reference:" << std::endl;
    std::cout << to_str("a_copy: ") << to_str(a_copy) << std::endl;
    auto a[0] = 999;
    std::cout << to_str("\\nAfter modifying original a[0] = 999:") << std::endl;
    std::cout << to_str("original a: ") << to_str(a) << std::endl;
    std::cout << to_str("copied a_copy: ") << to_str(a_copy) << std::endl;
}


void test_pickle_serialization() {
    bool __fluxus_exc=false;
    /* docstring */
    std::cout << "\n=== Testing Pickle Serialization ===" << std::endl;
    auto data = {;
        'string': 'hello world',;
        'list': std::vector<decltype(1)>{1, 2.5, true, 0},;
        'dict': {'nested': {'deep': 'value'}},;
        'tuple': (1, 2, 3),;
        'set': {4, 5, 6};
    }
    };
    std::cout << "Original data:" << " " << data << std::endl;
    auto pickled = pickle.dumps(data);
    std::cout << to_str("Pickled data size: ") << to_str(len(pickled)) << to_str(" bytes") << std::endl;
    auto unpickled = pickle.loads(pickled);
    std::cout << "Unpickled data:" << " " << unpickled << std::endl;
    std::cout << "Data equality:" << " " << data == unpickled << std::endl;
    for (int protocol = 0; protocol < pickle.HIGHEST_PROTOCOL + 1; ++protocol) {
        {
            auto p = pickle.dumps(data, protocol=protocol);
            auto u = pickle.loads(p);
            std::cout << to_str("Protocol ") << to_str(protocol) << to_str(": size = ") << to_str(len(p)) << to_str(" bytes, works = ") << to_str(data == u) << std::endl;
        }
        if (__fluxus_exc) { auto e = 0;
            std::cout << to_str("Protocol ") << to_str(protocol) << to_str(": failed - ") << to_str(e) << std::endl;
        }
    }
}


void test_shelve_persistence() {
    bool __fluxus_exc=false;
    /* docstring */
    std::cout << "\n=== Testing Shelve Persistence ===" << std::endl;
    auto temp_dir = tempfile.mkdtemp();
    auto db_path = os.path.join(temp_dir, 'test_shelve');
    {
        {
            auto db['user_data'] = std::unordered_map<std::string, std::string>{{"name", 'Alice'}, {"age", to_str(30)}};
            auto db['settings'] = std::unordered_map<std::string, std::string>{{"theme", 'dark'}, {"language", 'en'}};
            auto db['numbers'] = std::vector<decltype(1)>{1, 2, 3, 4, 5};
            auto db['timestamp'] = time.time();
            std::cout << "Data written to shelve database" << std::endl;
        }
        {
            std::cout << "Data read from shelve database:" << std::endl;
            for (auto [key, value] : py_items(db)) {
                std::cout << to_str("  ") << to_str(key) << to_str(": ") << to_str(value) << std::endl;
            }
        }
        {
            auto user_data = db['user_data'];
            auto user_data['age'] = 31;
            auto db['user_data'] = user_data;
            auto db['new_key'] = 'new_value';
            std::cout << "\nData modified in database" << std::endl;
        }
        {
            std::cout << "Modified data:" << std::endl;
            std::cout << to_str("  user_data: ") << to_str(db['user_data']) << std::endl;
            std::cout << to_str("  new_key: ") << to_str(db['new_key']) << std::endl;
        }
    }
    {
        for (auto file_path : glob.glob(f"{db_path}*")) {
            {
                os.unlink(file_path);
            }
            if (__fluxus_exc) { auto e = 0;
                /* pass */
            }
        }
        {
            os.rmdir(temp_dir);
        }
        if (__fluxus_exc) { auto e = 0;
            /* pass */
        }
    }
}


void test_copy_performance() {
    /* docstring */
    std::cout << "\n=== Testing Copy Performance ===" << std::endl;
    auto large_data = {;
        0;
    }
    };
    auto start = time.time();
    auto shallow_copy = copy.copy(large_data);
    auto shallow_time = time.time() - start;
    auto start = time.time();
    auto deep_copy = copy.deepcopy(large_data);
    auto deep_time = time.time() - start;
    std::cout << to_str("Shallow copy time: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(4); os<<shallow_time; return os.str(); }()) << to_str(" seconds") << std::endl;
    std::cout << to_str("Deep copy time: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(4); os<<deep_time; return os.str(); }()) << to_str(" seconds") << std::endl;
    std::cout << to_str("Deep copy is ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1); os<<deep_time/shallow_time; return os.str(); }()) << to_str("x slower") << std::endl;
    std::cout << to_str("Shallow copy first item data == original: ") << to_str(shallow_copy['items'][0]['data'] == large_data['items'][0]['data']) << std::endl;
    std::cout << to_str("Deep copy first item data == original: ") << to_str(deep_copy['items'][0]['data'] == large_data['items'][0]['data']) << std::endl;
    auto large_data['items'][0]['data'][0] = 99999;
    std::cout << to_str("After modifying original, shallow copy reflects change: ") << to_str(shallow_copy['items'][0]['data'][0] == 99999) << std::endl;
    std::cout << to_str("After modifying original, deep copy remains unchanged: ") << to_str(deep_copy['items'][0]['data'][0] == 0) << std::endl;
}


int main() {
    /* docstring */
    测试Python的深拷贝vs浅拷贝行为;
    /* docstring */
    test_shallow_vs_deep_copy();
    test_copy_edge_cases();
    test_pickle_serialization();
    test_shelve_persistence();
    test_copy_performance();
    std::cout << "\n=== All copy and serialization tests completed ===" << std::endl;
    return 0;
}

